#  David Vaughan  R1166390
# (MATH-4000-004)
# Homework 1 01/25/2023
#  Spring 2023 TTU Selected Topics: Machine learning model order reduction for differential equations (MATH-4000-004) Full Term: Homework for week 1 (due 01/25)

### Problem 1
Find out all the integers between 1 and 10000, which satisfies that this integer plus 268 is a square number.

In [75]:
### num is 268
# list stores int + num that are square number
### Square numbers can also be called perfect square numbers.
# a square number or perfect square is an integer that is the square of an integer 
# in other words, it is the product of some integer with itself. 
#For example, 9 is a square number, since it equals 3 to the power 2 and can be written as 3 × 3. 

# Probably  more effecient way to do this function
# add int between 1 and 10000 with 256, square root it , then if it is a int then it is a square number and add it to a list
import math
def prob1(start,end,num,list):
    ### num is 268
    # python will end on 9999 if end is 10000 so add 1
    for i in range(start, end+1):
        x = i + num
        y = math.sqrt(x)
        # If y is a int/whole number then it will be from a perfect square. 
        if (y.is_integer()) == True:
            # print(i)
            list.append(i)
        
    print("done with function")


In [76]:
list1 =[]
prob1(1,10000,268,list1)

done with function


In [77]:
print(list1)
# 21 + 268 = 289 square root 17
#
# 9933 + 268 = 10201 square root 101
# probably could do something with this info

[21, 56, 93, 132, 173, 216, 261, 308, 357, 408, 461, 516, 573, 632, 693, 756, 821, 888, 957, 1028, 1101, 1176, 1253, 1332, 1413, 1496, 1581, 1668, 1757, 1848, 1941, 2036, 2133, 2232, 2333, 2436, 2541, 2648, 2757, 2868, 2981, 3096, 3213, 3332, 3453, 3576, 3701, 3828, 3957, 4088, 4221, 4356, 4493, 4632, 4773, 4916, 5061, 5208, 5357, 5508, 5661, 5816, 5973, 6132, 6293, 6456, 6621, 6788, 6957, 7128, 7301, 7476, 7653, 7832, 8013, 8196, 8381, 8568, 8757, 8948, 9141, 9336, 9533, 9732, 9933]


### Answer to Problem 1
21, 56, 93, 132, 173, 216, 261, 308, 357, 408, 461, 516, 573, 632, 693, 756, 821, 888, 957, 1028, 1101, 1176, 1253, 1332, 1413, 1496, 1581, 1668, 1757, 1848, 1941, 2036, 2133, 2232, 2333, 2436, 2541, 2648, 2757, 2868, 2981, 3096, 3213, 3332, 3453, 3576, 3701, 3828, 3957, 4088, 4221, 4356, 4493, 4632, 4773, 4916, 5061, 5208, 5357, 5508, 5661, 5816, 5973, 6132, 6293, 6456, 6621, 6788, 6957, 7128, 7301, 7476, 7653, 7832, 8013, 8196, 8381, 8568, 8757, 8948, 9141, 9336, 9533, 9732, 9933


### Problem 2
The factorial of any positive integer $k$ is defined as:
$$k! = k \times (k-1) \times \cdots \times 2 \times 1$$
Compute the following expression:
$$\sum_{k=1}^9 k!$$

In [78]:
###
# list not necessary
# use for loop and math factorial function to add them together
#  diferent ways to do this at bottom of file
import math
def prob2Fac(start,end,fac_list):
    sum = 0
    # python will end on 8 if end is 9 so add 1
    for i in range(start, end+1):
        # Use math module  to calculate factorial instead of list
        print ("k is ",i)
        fac = math.factorial(i)
        
        fac_list.append(fac)
        print("Factorial of k is",fac)
        sum += fac
    print("sum is",sum)
    print("done with function")
    return sum

In [79]:
factorial_list = []
sum = prob2Fac(1,9,factorial_list)

k is  1
Factorial of k is 1
k is  2
Factorial of k is 2
k is  3
Factorial of k is 6
k is  4
Factorial of k is 24
k is  5
Factorial of k is 120
k is  6
Factorial of k is 720
k is  7
Factorial of k is 5040
k is  8
Factorial of k is 40320
k is  9
Factorial of k is 362880
sum is 409113
done with function


In [80]:
print(sum)

409113


### Answer to Problem 2
409113


### Problem 3
Give a string consists of several letters. If the letter is in the even-position, convert it to lower case; if the letter is in the odd-position, convert it to upper case.
For example, if the given string is "Morning", the the result should be "MoRnInG".

Convert the string "Welcome" and print the result.

In [81]:
###
# 
# in Python Welcome is 0-6 with W at 0
# Assume for problem 1-7 with W at 1
# Go through the length of the word, check if even or odd.
# if odd convert to upper case and add that charachter it to a new string
# else if even convert it to lower case and add  to new string
def prob3(word):
    new_word = ""
    size = len(word)
    print ("len of word is",size)
    print("word at 0 is",word[0])
    print("word at 6 is",word[6])
    for i in range(0, size):
        
        x = i+1
        # print(x)
        # Check if even
        if(x%2) == 0:
            new_word = new_word+ word[i].lower() 
            #print("even")
        else:
            new_word = new_word+ word[i].upper() 
            #print("odd")
        
    print("done with function 3")
    return new_word

In [82]:
word = "Welcome"
word2 = prob3(word)

len of word is 7
word at 0 is W
word at 6 is e
done with function 3


In [83]:
print(word2)

WeLcOmE


### Answer to Problem 3
WeLcOmE


### Problem 4
Use for loop to print the following pattern:

^^^^= \
^^^== \
^^=== \
^==== \
=====

In [84]:
###
# 
def prob4c(length):
    x = length
    for i in range(length):
        # the ^ charachter should get smaller with ever line, x achieves that
        x -= 1
        for _ in range(x):
            print('^', end='')
        # gets bigger ever line
        for p in range(i+1):
            print('=', end='')
        print("")
    print("done with function")

In [85]:
prob4c(5)

^^^^=
^^^==
^^===
^====
=====
done with function


### Answer to Problem 4
^^^^= \
^^^== \
^^=== \
^==== \
=====


### Problem 5
In number theory, the Padovan sequence is the sequence of integers $P(n)$ defined by the initial values
$$P(0) = P(1) = P(2) = 1$$
and the recurrence relation
$$P(n) = P(n-2) + P(n-3)$$
For example, the first few values of $P(n)$ are
1, 1, 1, 2, 2, 3, 4, 5, 7, 9

Compute and print $P(20)$.

In [86]:
###
# use a list to keep track of previous numbers in the Padovan sequence. 
# Use previous numbers in list to calulate new values that are added to list
def prob5(end,pad_list):
    for i in range(3, end+1):
        x = pad_list[i-2] + pad_list[i-3]
        pad_list.append(x)
    print("done with function")

    
   

In [87]:
padov_list = [1,1,1]
prob5(20,padov_list)

done with function


In [88]:
print(padov_list)

[1, 1, 1, 2, 2, 3, 4, 5, 7, 9, 12, 16, 21, 28, 37, 49, 65, 86, 114, 151, 200]


### Answer to Problem 5
1, 1, 1, 2, 2, 3, 4, 5, 7, 9, 12, 16, 21, 28, 37, 49, 65, 86, 114, 151, 200

### END OF HOMEWORK
#
#
#
#
#

### Before you work on this homework, please self-study the following list:
1. Anaconda: a platform of Python packages for scientific computing (data science, machine learning applications, large-scale data processing, predictive analytics, etc.)
    - download and install Anaconda from https://www.anaconda.com/
    - install Python and Jupyter notebook in Anaconda
2. Python:
    - read and go through all the examples in this website: https://learnxinyminutes.com/docs/python/
    - learn how to run Python code in your computer
2. Jupyter:
    - learn how to run Jupyter Notebook
3. Markdown: a plain text markup language
    - read and go through all the examples in this website: https://learnxinyminutes.com/docs/markdown/

### Please first read the examples and then work on the following homework (Problem 1-5). Submit your solution code in the format of Jupyter Notebook as in the examples.

### Example 1:
Compute and print the summation of all the integers from 1 to 100:
$$\sum_{k=1}^{100} k$$
#### Solution:

In [89]:
sum = 0
for x in range(1, 101):
    #print(x)
    sum += x
print(sum)

5050


### Example 2:
Use for loop to print the following pattern:

= \
== \
=== \
==== \
=====
#### Solution:

In [90]:
row = 5
for i in range(row):
    for _ in range(i + 1):
        print('=', end='')
    print("l")

=l
==l
===l
====l
=====l


In [91]:
###
# list not necessary
def prob2Fac(start,end,factorial_previous):
    sum = 0
    # python will end on 8 if end is 9 so add 1
    for i in range(start, end+1):
        # Use math module  to calculate factorial instead of list
        print ("k is ",i)
        factorial_previous = i*factorial_previous
        print("Factorial of k is",factorial_previous)
        sum += factorial_previous
    print("sum is",sum)
    print("done with function")
    return sum

In [92]:
#
sum = prob2Fac(1,9,1)

k is  1
Factorial of k is 1
k is  2
Factorial of k is 2
k is  3
Factorial of k is 6
k is  4
Factorial of k is 24
k is  5
Factorial of k is 120
k is  6
Factorial of k is 720
k is  7
Factorial of k is 5040
k is  8
Factorial of k is 40320
k is  9
Factorial of k is 362880
sum is 409113
done with function


In [93]:
###
# 
def prob4(length):
    x = length
    y = 0
    for i in range(length):
        y +=1
        x -= 1
        for _ in range(x):
            print('^', end='')
        for w in range(y):
            print('=', end='')
        print("")
    print("done with function")

In [94]:
prob4(5)

^^^^=
^^^==
^^===
^====
=====
done with function


In [95]:
# online different ways to replace Strings
# Ways to replace elements in string
# Method 1: replace character in string by index
def replace_char1(str, index, char):
    str = list(str)
    str[index] = char
    return ''.join(str)

# method 2: using string slicing
def replace_char2(str, index, char):
    str = str[:index] + char + str[index+1:]
    return str

In [96]:
# Online Python program to check even number
# using sympy.is_even() method
# importing sympy module
# from sympy import 
# calling is_evenfunction on different numbers
#  geek1 = simplify(30).is_even

In [97]:
import math
x = math.sqrt(9)
print(x.is_integer())
y= math.sqrt(8)
print(y.is_integer())

True
False
